File: space/hobby_image/dit.py
"""DreamLite in-context DiT on a DC-AE f32c32 latent — the $300 path.

DC-AE compresses 32x spatially (vs VAE f8's 8x), so a 512px image -> 16x16x32 latent. The two-panel
canvas (target|source, width-concat) is 16x32x32 = ONLY 512 tokens (vs the conv U-Net's 64x128=8,192
positions). A DiT on 512 tokens is ~order-of-magnitude cheaper to train than the conv U-Net, which is
how the budget drops from ~$1.66k toward ~$300. Trade-off: f32 reconstruction is slightly weaker than
f8c16 on fine text (acceptable for the V0 "global + simple-local edits" scope).

Standard DiT (AdaLN-Zero on self-attn + MLP) + cross-attention to the frozen-VLM refiner tokens, flow-
matching velocity. Two-panel handled by a learned left/right panel embedding; loss uses the left half.
"""
from __future__ import annotations

import math
from dataclasses import dataclass

import torch
import torch.nn as nn
import torch.nn.functional as F
from torch import Tensor


@dataclass
class DiTConfig:
    in_channels: int = 34      # 32 DC-AE latent ch + 2 optional mask ch
    out_channels: int = 32
    latent_h: int = 16         # 512px / 32 = 16 (256px -> 8)
    panel_w: int = 16          # per-panel width; canvas width = 2*panel_w
    patch: int = 1             # DC-AE latent is already compressed -> patch 1
    d_model: int = 1024
    depth: int = 16
    heads: int = 16
    mlp_ratio: float = 3.0
    ctx_dim: int = 1024
    n_tasks: int = 3


def sinusoidal(t: Tensor, dim: int) -> Tensor:
    half = dim // 2
    f = torch.exp(-math.log(10000.0) * torch.arange(half, device=t.device) / half)
    a = t.float()[:, None] * f[None, :]
    return torch.cat([a.cos(), a.sin()], dim=-1)


def modulate(x: Tensor, shift: Tensor, scale: Tensor) -> Tensor:
    return x * (1 + scale[:, None]) + shift[:, None]


class RMSNorm(nn.Module):
    def __init__(self, dim: int, eps: float = 1e-6):
        super().__init__()
        self.weight = nn.Parameter(torch.ones(dim))
        self.eps = eps

    def forward(self, x: Tensor) -> Tensor:
        return F.rms_norm(x, (x.size(-1),), self.weight, self.eps)


class Attention(nn.Module):
    def __init__(self, dim: int, heads: int, ctx_dim: int | None = None, qk_norm: bool = True):
        super().__init__()
        self.heads = heads
        self.hd = dim // heads
        self.cross = ctx_dim is not None
        self.q = nn.Linear(dim, dim, bias=False)
        self.kv = nn.Linear(ctx_dim or dim, 2 * dim, bias=False)
        self.o = nn.Linear(dim, dim, bias=False)
        # QK-norm: RMSNorm on Q,K (over head_dim) BEFORE the dot product — the key training-stability
        # fix for DiTs (Inf-DiT/Lumina-Next/SD3); bounds the attention logits, prevents divergence.
        self.qn = RMSNorm(self.hd) if qk_norm else nn.Identity()
        self.kn = RMSNorm(self.hd) if qk_norm else nn.Identity()

    def forward(self, x: Tensor, ctx: Tensor | None = None) -> Tensor:
        B, N, _ = x.shape
        src = ctx if self.cross else x
        q = self.qn(self.q(x).view(B, N, self.heads, self.hd)).transpose(1, 2)
        kv = self.kv(src).view(B, src.shape[1], 2, self.heads, self.hd)
        k = self.kn(kv[:, :, 0]).transpose(1, 2)
        v = kv[:, :, 1].transpose(1, 2)
        o = F.scaled_dot_product_attention(q, k, v)
        return self.o(o.transpose(1, 2).reshape(B, N, -1))


class DiTBlock(nn.Module):
    def __init__(self, cfg: DiTConfig):
        super().__init__()
        d = cfg.d_model
        self.ln1 = nn.LayerNorm(d, elementwise_affine=False, eps=1e-6)
        self.sa = Attention(d, cfg.heads)
        self.lnc = nn.LayerNorm(d, eps=1e-6)
        self.ca = Attention(d, cfg.heads, ctx_dim=cfg.ctx_dim)
        self.ln2 = nn.LayerNorm(d, elementwise_affine=False, eps=1e-6)
        h = int(d * cfg.mlp_ratio)
        self.mlp = nn.Sequential(nn.Linear(d, h), nn.GELU(approximate="tanh"), nn.Linear(h, d))
        self.adaln = nn.Sequential(nn.SiLU(), nn.Linear(d, 6 * d))
        nn.init.zeros_(self.adaln[-1].weight); nn.init.zeros_(self.adaln[-1].bias)

    def forward(self, x: Tensor, cond: Tensor, ctx: Tensor) -> Tensor:
        sa_s, sa_sc, sa_g, ml_s, ml_sc, ml_g = self.adaln(cond).chunk(6, dim=-1)
        x = x + sa_g[:, None] * self.sa(modulate(self.ln1(x), sa_s, sa_sc))
        x = x + self.ca(self.lnc(x), ctx)                       # cross-attn to VLM tokens
        x = x + ml_g[:, None] * self.mlp(modulate(self.ln2(x), ml_s, ml_sc))
        return x


class HobbyImageDiT(nn.Module):
    def __init__(self, cfg: DiTConfig):
        super().__init__()
        self.cfg = cfg
        d = cfg.d_model
        self.canvas_w = 2 * cfg.panel_w
        self.n_tokens = (cfg.latent_h // cfg.patch) * (self.canvas_w // cfg.patch)
        self.patch_embed = nn.Conv2d(cfg.in_channels, d, cfg.patch, stride=cfg.patch)
        self.pos = nn.Parameter(torch.zeros(1, self.n_tokens, d))
        self.panel = nn.Parameter(torch.zeros(2, d))            # left/right
        self.t_mlp = nn.Sequential(nn.Linear(d, d), nn.SiLU(), nn.Linear(d, d))
        self.task_emb = nn.Embedding(cfg.n_tasks, d)
        self.blocks = nn.ModuleList([DiTBlock(cfg) for _ in range(cfg.depth)])
        self.ln_f = nn.LayerNorm(d, elementwise_affine=False, eps=1e-6)
        self.adaln_f = nn.Sequential(nn.SiLU(), nn.Linear(d, 2 * d))
        nn.init.zeros_(self.adaln_f[-1].weight); nn.init.zeros_(self.adaln_f[-1].bias)
        self.head = nn.Linear(d, cfg.patch * cfg.patch * cfg.out_channels)
        nn.init.zeros_(self.head.weight); nn.init.zeros_(self.head.bias)
        nn.init.trunc_normal_(self.pos, std=0.02)

    def forward(self, x: Tensor, t: Tensor, ctx: Tensor, task: Tensor | None = None) -> Tensor:
        """x: (B, in_ch, H, 2*panel_w) two-panel DC-AE latent. t: (B,). ctx: (B, M, ctx_dim).
        Returns velocity (B, out_ch, H, 2*panel_w)."""
        cfg = self.cfg
        B, _, H, W = x.shape
        h = self.patch_embed(x).flatten(2).transpose(1, 2)      # (B, N, d)
        h = h + self.pos
        # panel embedding: tokens in the left half vs right half (by column)
        gw = W // cfg.patch
        col = torch.arange(self.n_tokens, device=x.device) % gw
        left = (col < gw // 2)
        h = h + torch.where(left[None, :, None], self.panel[0], self.panel[1])
        cond = self.t_mlp(sinusoidal(t * 1000.0, cfg.d_model))
        if task is not None:
            cond = cond + self.task_emb(task)
        for blk in self.blocks:
            h = blk(h, cond, ctx)
        s, sc = self.adaln_f(cond).chunk(2, dim=-1)
        h = modulate(self.ln_f(h), s, sc)
        h = self.head(h)                                        # (B, N, p*p*out)
        # unpatchify
        gh = H // cfg.patch
        h = h.view(B, gh, gw, cfg.patch, cfg.patch, cfg.out_channels)
        h = h.permute(0, 5, 1, 3, 2, 4).reshape(B, cfg.out_channels, H, W)
        return h


def count_params(m: nn.Module) -> int:
    return sum(p.numel() for p in m.parameters())


# 512px: DC-AE f32 -> 16x16x32 per panel; two-panel canvas 16x32; 512 tokens.
V0_DCAE_512 = DiTConfig(in_channels=34, out_channels=32, latent_h=16, panel_w=16, patch=1,
                        d_model=1024, depth=16, heads=16, mlp_ratio=3.0, ctx_dim=1024)
# 256px pilot: 8x8x32 per panel; canvas 8x16; 128 tokens.
V0_DCAE_256 = DiTConfig(in_channels=34, out_channels=32, latent_h=8, panel_w=8, patch=1,
                        d_model=1024, depth=16, heads=16, mlp_ratio=3.0, ctx_dim=1024)


if __name__ == "__main__":
    for name, cfg in [("512", V0_DCAE_512), ("256", V0_DCAE_256)]:
        m = HobbyImageDiT(cfg)
        x = torch.randn(2, cfg.in_channels, cfg.latent_h, 2 * cfg.panel_w)
        t = torch.rand(2); ctx = torch.randn(2, 256, cfg.ctx_dim); task = torch.zeros(2, dtype=torch.long)
        with torch.no_grad():
            y = m(x, t, ctx, task)
        print(f"DiT {name}: {count_params(m)/1e6:.1f}M params | {m.n_tokens} tokens | "
              f"in {tuple(x.shape)} -> out {tuple(y.shape)}")

File: hobby-rs/src/dit.rs
//! HobbyImageDiT: the in-context flow-matching DiT (333M) that denoises the DC-AE latent, conditioned
//! on CLIP text tokens via cross-attention. Port of hobby_image/dit.py. Two-panel canvas (target|source)
//! with AdaLN-Zero self-attention + cross-attention + GELU-tanh MLP; outputs a velocity field.

use crate::imgops::{conv2d, gelu_tanh, linear, rmsnorm_rows, Map};
use crate::st::SafeTensors;
use rayon::prelude::*;
use std::path::Path;

pub struct DiTConfig {
    pub in_channels: usize, // 34
    pub out_channels: usize, // 32
    pub latent_h: usize,    // 32 -> 1024px
    pub panel_w: usize,     // 32
    pub d_model: usize,     // 1024
    pub depth: usize,       // 16
    pub heads: usize,       // 16
    pub ctx_dim: usize,     // 768
}

impl Default for DiTConfig {
    fn default() -> Self {
        DiTConfig { in_channels: 34, out_channels: 32, latent_h: 32, panel_w: 32,
                    d_model: 1024, depth: 16, heads: 16, ctx_dim: 768 }
    }
}

pub struct Dit {
    st: SafeTensors,
    pub cfg: DiTConfig,
}

/// LayerNorm with NO affine (ln1/ln2/ln_f): subtract mean, divide by std.
fn ln_noaffine(x: &[f32], n: usize, d: usize, eps: f32) -> Vec<f32> {
    let mut out = vec![0.0f32; n * d];
    out.par_chunks_mut(d).enumerate().for_each(|(i, o)| {
        let row = &x[i * d..(i + 1) * d];
        let mean = row.iter().sum::<f32>() / d as f32;
        let var = row.iter().map(|v| (v - mean) * (v - mean)).sum::<f32>() / d as f32;
        let inv = 1.0 / (var + eps).sqrt();
        for j in 0..d {
            o[j] = (row[j] - mean) * inv;
        }
    });
    out
}

/// modulate: x * (1 + scale) + shift, broadcasting (d,) over n rows.
fn modulate(x: &[f32], n: usize, d: usize, shift: &[f32], scale: &[f32]) -> Vec<f32> {
    let mut out = vec![0.0f32; n * d];
    for i in 0..n {
        for j in 0..d {
            out[i * d + j] = x[i * d + j] * (1.0 + scale[j]) + shift[j];
        }
    }
    out
}

fn sinusoidal(t: f32, dim: usize) -> Vec<f32> {
    let half = dim / 2;
    let mut out = vec![0.0f32; dim];
    for i in 0..half {
        let f = (-(10000f32.ln()) * i as f32 / half as f32).exp();
        let a = t * f;
        out[i] = a.cos();
        out[half + i] = a.sin();
    }
    out
}

impl Dit {
    pub fn load(dir: &Path, cfg: DiTConfig) -> anyhow::Result<Dit> {
        let st = SafeTensors::open(&dir.join("dit.safetensors"))?;
        Ok(Dit { st, cfg })
    }

    /// Multi-head attention. q from `xq` (nq rows), k/v from `xkv` (nk rows). Per-head RMSNorm on
    /// q and k (weights qn/kn over head_dim) before the dot product. No causal mask (full attention).
    #[allow(clippy::too_many_arguments)]
    fn attention(&self, prefix: &str, xq: &[f32], nq: usize, xkv: &[f32], nk: usize,
                 kv_in: usize) -> Vec<f32> {
        let d = self.cfg.d_model;
        let heads = self.cfg.heads;
        let hd = d / heads;
        let g = |n: &str| self.st.data(n);
        let q = linear(xq, nq, d, g(&format!("{prefix}.q.weight")), None, d);
        let kv = linear(xkv, nk, kv_in, g(&format!("{prefix}.kv.weight")), None, 2 * d);
        // split kv -> k (col 0..d), v (col d..2d), with the view (nk, 2, heads, hd)
        let mut k = vec![0.0f32; nk * d];
        let mut v = vec![0.0f32; nk * d];
        for i in 0..nk {
            k[i * d..(i + 1) * d].copy_from_slice(&kv[i * 2 * d..i * 2 * d + d]);
            v[i * d..(i + 1) * d].copy_from_slice(&kv[i * 2 * d + d..i * 2 * d + 2 * d]);
        }
        // per-head RMSNorm on q and k over head_dim (reshape rows to nq*heads x hd)
        let qn = g(&format!("{prefix}.qn.weight"));
        let kn = g(&format!("{prefix}.kn.weight"));
        let q = rmsnorm_rows(&q, nq * heads, hd, qn, 1e-6);
        let k = rmsnorm_rows(&k, nk * heads, hd, kn, 1e-6);
        let scale = 1.0 / (hd as f32).sqrt();
        // output (nq, d), parallel over (head, query row)
        let mut out = vec![0.0f32; nq * d];
        let cells: Vec<(usize, usize)> = (0..heads).flat_map(|h| (0..nq).map(move |i| (h, i))).collect();
        let results: Vec<(usize, usize, Vec<f32>)> = cells
            .par_iter()
            .map(|&(h, i)| {
                let off = h * hd;
                let qslice = &q[(i * heads + h) * hd..(i * heads + h) * hd + hd];
                let mut scores = vec![0.0f32; nk];
                for j in 0..nk {
                    let kslice = &k[(j * heads + h) * hd..(j * heads + h) * hd + hd];
                    scores[j] = crate::ops::dot(qslice, kslice) * scale;
                }
                let m = scores.iter().cloned().fold(f32::NEG_INFINITY, f32::max);
                let mut sum = 0.0f32;
                for s in scores.iter_mut() {
                    *s = (*s - m).exp();
                    sum += *s;
                }
                let inv = 1.0 / sum;
                let mut o = vec![0.0f32; hd];
                // weighted sum of value rows (contiguous head slice) — vectorized axpy over hd.
                let nlanes = hd - hd % 8;
                for j in 0..nk {
                    let wj = scores[j] * inv;
                    let wv = wide::f32x8::splat(wj);
                    let vrow = &v[j * d + off..j * d + off + hd];
                    let mut t = 0;
                    while t < nlanes {
                        let ov = wide::f32x8::new(<[f32; 8]>::try_from(&o[t..t + 8]).unwrap());
                        let vv = wide::f32x8::new(<[f32; 8]>::try_from(&vrow[t..t + 8]).unwrap());
                        let r = wv.mul_add(vv, ov);
                        o[t..t + 8].copy_from_slice(r.as_array_ref());
                        t += 8;
                    }
                    while t < hd {
                        o[t] += wj * vrow[t];
                        t += 1;
                    }
                }
                (h, i, o)
            })
            .collect();
        for (h, i, o) in results {
            let off = h * hd;
            out[i * d + off..i * d + off + hd].copy_from_slice(&o);
        }
        linear(&out, nq, d, g(&format!("{prefix}.o.weight")), None, d)
    }

    /// Forward one denoising eval. `x` is the (34, H, 2*panel_w) two-panel input; `t` scalar in [0,1);
    /// `ctx` is the CLIP last_hidden_state (m x ctx_dim); `task` the task id. Returns the velocity
    /// field as a Map(out_channels, H, 2*panel_w) — caller slices the left panel.
    pub fn forward(&self, x: &Map, t: f32, ctx: &[f32], m: usize, task: usize) -> Map {
        let cfg = &self.cfg;
        let d = cfg.d_model;
        let h_lat = cfg.latent_h;
        let w = 2 * cfg.panel_w;
        let n = h_lat * w; // tokens
        let g = |nm: &str| self.st.data(nm);

        // patch embed (1x1 conv) -> (d, H, W); transpose to tokens (n, d) with t = y*W + x
        let pe = conv2d(x, g("patch_embed.weight"), Some(g("patch_embed.bias")), d, 1, 0, 1);
        let mut h = vec![0.0f32; n * d];
        for c in 0..d {
            for y in 0..h_lat {
                for xx in 0..w {
                    h[(y * w + xx) * d + c] = pe.d[(c * h_lat + y) * w + xx];
                }
            }
        }
        // + positional embedding (1, n, d)
        let pos = g("pos");
        for i in 0..n * d {
            h[i] += pos[i];
        }
        // + panel embedding: left half (col < panel_w) vs right
        let panel = g("panel"); // (2, d)
        for tkn in 0..n {
            let col = tkn % w;
            let p = if col < cfg.panel_w { 0 } else { 1 };
            for c in 0..d {
                h[tkn * d + c] += panel[p * d + c];
            }
        }
        // conditioning vector: t_mlp(sinusoidal(t*1000)) + task_emb(task)
        let sin = sinusoidal(t * 1000.0, d);
        let mut cond = linear(&sin, 1, d, g("t_mlp.0.weight"), Some(g("t_mlp.0.bias")), d);
        for v in cond.iter_mut() {
            *v = crate::imgops::silu(*v);
        }
        cond = linear(&cond, 1, d, g("t_mlp.2.weight"), Some(g("t_mlp.2.bias")), d);
        let task_emb = g("task_emb.weight");
        for c in 0..d {
            cond[c] += task_emb[task * d + c];
        }

        for l in 0..cfg.depth {
            let p = format!("blocks.{l}");
            // adaLN: SiLU(cond) then Linear -> 6d
            let mut cs = cond.clone();
            for v in cs.iter_mut() {
                *v = crate::imgops::silu(*v);
            }
            let mod6 = linear(&cs, 1, d, g(&format!("{p}.adaln.1.weight")),
                              Some(g(&format!("{p}.adaln.1.bias"))), 6 * d);
            let (sa_s, sa_sc, sa_g) = (&mod6[0..d], &mod6[d..2 * d], &mod6[2 * d..3 * d]);
            let (ml_s, ml_sc, ml_g) = (&mod6[3 * d..4 * d], &mod6[4 * d..5 * d], &mod6[5 * d..6 * d]);

            // self-attention with AdaLN-Zero gate
            let ln1 = ln_noaffine(&h, n, d, 1e-6);
            let modx = modulate(&ln1, n, d, sa_s, sa_sc);
            let sa = self.attention(&format!("{p}.sa"), &modx, n, &modx, n, d);
            for i in 0..n {
                for c in 0..d {
                    h[i * d + c] += sa_g[c] * sa[i * d + c];
                }
            }
            // cross-attention to ctx (lnc is affine LayerNorm), no gate
            let lnc = crate::imgops::layernorm(&h, n, d, g(&format!("{p}.lnc.weight")),
                                               g(&format!("{p}.lnc.bias")), 1e-6);
            let ca = self.attention(&format!("{p}.ca"), &lnc, n, ctx, m, cfg.ctx_dim);
            for i in 0..n * d {
                h[i] += ca[i];
            }
            // MLP with AdaLN-Zero gate
            let ln2 = ln_noaffine(&h, n, d, 1e-6);
            let modx = modulate(&ln2, n, d, ml_s, ml_sc);
            let hidden = g(&format!("{p}.mlp.0.weight")).len() / d;
            let mut f1 = linear(&modx, n, d, g(&format!("{p}.mlp.0.weight")),
                                Some(g(&format!("{p}.mlp.0.bias"))), hidden);
            for v in f1.iter_mut() {
                *v = gelu_tanh(*v);
            }
            let f2 = linear(&f1, n, hidden, g(&format!("{p}.mlp.2.weight")),
                            Some(g(&format!("{p}.mlp.2.bias"))), d);
            for i in 0..n {
                for c in 0..d {
                    h[i * d + c] += ml_g[c] * f2[i * d + c];
                }
            }
        }
        // final AdaLN (2d) + head
        let mut cs = cond.clone();
        for v in cs.iter_mut() {
            *v = crate::imgops::silu(*v);
        }
        let mod2 = linear(&cs, 1, d, g("adaln_f.1.weight"), Some(g("adaln_f.1.bias")), 2 * d);
        let lnf = ln_noaffine(&h, n, d, 1e-6);
        let hmod = modulate(&lnf, n, d, &mod2[0..d], &mod2[d..2 * d]);
        let out = linear(&hmod, n, d, g("head.weight"), Some(g("head.bias")), cfg.out_channels);
        // unpatchify (n, out_ch) -> Map(out_ch, H, W), token t = y*W + x
        let oc = cfg.out_channels;
        let mut map = Map::zeros(oc, h_lat, w);
        for y in 0..h_lat {
            for xx in 0..w {
                let tkn = y * w + xx;
                for c in 0..oc {
                    map.d[(c * h_lat + y) * w + xx] = out[tkn * oc + c];
                }
            }
        }
        map
    }
}
